torch/onnx/_internal/exporter.py

```diff
@@ -0,0 +1,230 @@
+# necessary to surface onnx.ModelProto through ExportOutput:
+from __future__ import annotations
+
+import abc
+import io
+import logging
+from typing import (
+    Any,
+    Callable,
+    Final,
+    Mapping,
+    Optional,
+    Protocol,
+    runtime_checkable,
+    Sequence,
+    TYPE_CHECKING,
+    TypeVar,
+    Union,
+)
+
+import torch
+
+from torch.onnx._internal import _beartype
+
+# We can only import onnx from this module in a type-checking context to ensure that
+# 'import torch.onnx' continues to work without having 'onnx' installed. We fully
+# 'import onnx' inside of dynamo_export (by way of _assert_dependencies).
+if TYPE_CHECKING:
+    import onnx
+
+_DEFAULT_OPSET_VERSION: Final[int] = 18
+
+
+@_beartype.beartype
+class ExportOptions:
+    """Options to influence the TorchDynamo ONNX exporter."""
+
+    opset_version: Optional[int] = None
+    """The ONNX opset version the exporter should target. Defaults to the latest
+    supported ONNX opset version. The default version will increment over time as
+    ONNX continues to evolve."""
+
+    dynamic_shapes: Optional[bool] = None
+    """Shape information hint for input/output tensors.
+
+    - ``None``: the exporter determines the most compatible setting.
+    - ``True``: all input shapes are considered dynamic.
+    - ``False``: all input shapes are considered static."""
+
+    logger: Optional[logging.Logger] = None
+    """The logger for the ONNX exporter to use. Defaults to creating a child
+    logger named "torch.onnx" under the current logger (as returned by
+    :py:meth:`logging.getLogger`)."""
+
+    def __init__(
+        self,
+        *,
+        opset_version: Optional[int] = None,
+        dynamic_shapes: Optional[bool] = None,
+        logger: Optional[logging.Logger] = None,
+    ):
+        self.opset_version = opset_version
+        self.dynamic_shapes = dynamic_shapes
+        self.logger = logger
+
+
+@runtime_checkable
+class ExportOutputSerializer(Protocol):
+    """Protocol for serializing an ONNX graph into a specific format (e.g. Protobuf)."""
+
+    def serialize(
+        self, export_output: ExportOutput, destination: io.BufferedIOBase
```

Comment:
`from typing import IO`

```suggestion
        self, export_output: ExportOutput, destination: IO[bytes]
```

Reply:
Beartype complains:

```
FAILED test/onnx/dynamo/test_exporter_api.py::TestDynamoExportAPI::test_save_to_existing_buffer_default_serializer - beartype.roar.BeartypeCallHintParamViolation: @beartyped torch.onnx._internal.expor...
```

Reply:
TL;DR seems a messy problem of python. `typing.IO` as a protocol defines too many abstract methods, which a lot of them do not exist for `io.BytesIO`.

Reply:
Yeah, my digging found a mess as well. Leaving this as `io.BufferedIOBase` for now. We can specialize more IO cases as we go if we need to.